// gulpex/lib/index.js
'use strict';

var assert = require('assert'),
    util = require('gulp-util'),
    fixExitStatus = require('./fix-exit-status'),
    catchSilentTaskFailure = require('./catch-silent-task-failure'),
    autoClean = require('./auto-clean'),
    run = require('./run'),
    makeWarner = require('./warn'),
    checkStale = require('./check-stale'),
    checkDepLinks = require('./check-dep-links'),
    download = require('./download'),
    announceWaiting = require('./announce-waiting'),
    path = require('path');

function gulpex(gulp, _module) {
    _module = _module || module.parent;
    assert(gulp && gulp.tasks, 'need gulp');
    assert(_module && _module.filename, 'need _module');

    require('stackup');

    fixExitStatus(gulp);
    catchSilentTaskFailure(gulp);
    announceWaiting(gulp, util);

    var warn = makeWarner(gulp),
        gulpfile = _module.filename,
        projectDir = path.dirname(gulpfile);

    return {
        run: run({
            // insert node_modules/.bin like npm does with scripts
            extraPaths: [ path.join(projectDir, 'node_modules', '.bin') ],
        }),
        util: util,
        clean: autoClean,
        warn: warn,
        download: download,
        checkStale: function _checkStale(done) {
            checkStale(warn, projectDir, done);
        },
        checkDepLinks: function _checkDepLinks(done) {
            checkDepLinks(warn, projectDir, done);
        }
    };
}

module.exports = gulpex;

require('pkginfo')(module);
