// hapi-anode/lib/ask-server.js
'use strict';

var assert = require('assert'),
    nipple = require('nipple'),
    util = require('util'),
    _ = require('lodash');

/**
 * @class Requestor
 * @constructor
 * @desc Method object for requests to the authentication server
 *
 * [Hapi scheme]: https://github.com/spumko/hapi/blob/master/docs/Reference.md#serverauthschemename-scheme
 * @param log {function} - the log from the original request
 * @param desc {string} - the request description for log and error messages
 * @param desc {method} - the request method
 * @param desc {uri} - the request URI
 * @param options {object} - the request options (NOT the strategy options!)
 * @param callback {function} - the callback
 */

function Requestor(log, desc, method, uri, options, callback) {
    assert(this instanceof Requestor);
    assert(typeof(log) === 'function');
    assert(typeof(desc) === 'string');
    assert(typeof(method) === 'string');
    assert(typeof(uri) === 'string');
    assert(typeof(options) === 'object');
    assert(typeof(callback) === 'function');

    this.log = log;
    this.desc = desc;
    this.method = method;
    this.uri = uri;
    this.options = options;
    this.callback = callback;
}

function censorOptions(options) {
    var result = _.cloneDeep(options);

    if (result.headers && result.headers.Authorization) {
        result.headers.Authorization = '...';
    }

    if (result.payload) {
        result.payload = '...';
    }

    return result;
}

Requestor.prototype.request = function request() {
    this.log([ 'auth' ], {
        what: this.describe('requesting %s')
    });
    this.log([ 'auth', 'trace' ], {
        what: this.describe('requesting %s'),
        method: this.method,
        uri: this.uri,
        options: censorOptions(this.options)
    });
    nipple.request(this.method, this.uri, this.options, this.onResponse.bind(this));
};

Requestor.prototype.describe = function describe(fmt) {
    return util.format(fmt, this.desc);
};

Requestor.prototype.onResponse = function onResponse(err, res) {
    if (err) {
        this.log([ 'auth', 'error' ], {
            what: this.describe('%s request failed'),
            err: err
        });
        return this.callbackWithFailure();
    }

    res.socket.on('close', function reportSocketClosed() {
        this.log(['auth', 'trace', 'socket'], {
            what: this.describe('%s socket closed')
        });
    }.bind(this));

    if (res.statusCode !== 200) {
        this.log([ 'auth', 'error' ], {
            what: this.describe('%s response status bad'),
            statusCode: res.statusCode
        });
        return this.callbackWithFailure();
    }

    nipple.read(res, this.onResponseBody.bind(this));
};

Requestor.prototype.callbackWithFailure = function callbackWithFailure() {
    var err = new Error(this.describe('%s request failed'));
    return this.callback(err);
};

Requestor.prototype.onResponseBody = function onResponseBody(err, body) {
    var parsed;

    if (err) {
        this.log([ 'auth', 'error' ], {
            what: this.describe('%s read failed'),
            err: err
        });
        return this.callbackWithFailure();
    }

    try {
        parsed = JSON.parse(body);
    } catch (parseErr) {
        this.log([ 'auth', 'error' ], {
            what: this.describe('%s response is not valid JSON'),
        });
        return this.callbackWithFailure();
    }

    this.callback(null, parsed);
};

function request(log, desc, method, uri, options, callback) {
    var worker = new Requestor(log, desc, method, uri, options, callback);
    return worker.request();
}

module.exports = request;
